pkg/synchronizer/synchronizer.go

```diff
@@ -0,0 +1,314 @@
+)
+
+type Synchronizer struct {
+	retryCount    uint64
+	retryInterval time.Duration
+	httpURLRoot   string
+
+	expectedCollectorCount int
+
+	ftpServer string
+	ftpUser   string
+	ftpPass   string
+
+	bh *storage.BucketHandle
+	gc pb.RVClient
+}
+
+type Config struct {
+	FTPServer string
+	FTPUser   string
+	FTPPass   string
+
+	GCSCli          *storage.Client
+	UploadServerCli pb.RVClient
+	ArchiveBucket   string
+	HTTPRoot        string
+}
+
+func New(cfg *Config) (*Synchronizer, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("nil config")
+	}
+	if cfg.FTPServer == "" || cfg.FTPUser == "" {
+		return nil, fmt.Errorf("invalid FTP creds")
+	}
+	if cfg.GCSCli == nil {
+		return nil, fmt.Errorf("invalid GCS client")
+	}
+	if cfg.UploadServerCli == nil {
+		return nil, fmt.Errorf("invalid gRPC connection")
+	}
+	if cfg.ArchiveBucket == "" {
+		return nil, fmt.Errorf("the corresponding archive bucket is not specified")
+	}
+	hr := cfg.HTTPRoot
+	if !strings.HasPrefix(hr, "http") {
+		return nil, fmt.Errorf("%s does not have the HTTP prefix", hr)
+	}
+	if hr[len(hr)-1] != '/' {
+		hr = hr + "/"
+	}
+
+	return &Synchronizer{
+		ftpServer: cfg.FTPServer,
+		ftpUser:   cfg.FTPUser,
+		ftpPass:   cfg.FTPPass,
+
+		gc: cfg.UploadServerCli,
+		bh: cfg.GCSCli.Bucket(cfg.ArchiveBucket),
+
+		// TODO: Hardcoded value. Parameterize this with a config.
+		expectedCollectorCount: 34,
+
+		retryCount:    5,
+		retryInterval: 10 * time.Second,
+		httpURLRoot:   hr,
+	}, nil
+}
+
+func (s *Synchronizer) uploadFromFTP(ctx context.Context, f string) (bool, error) {
+	gcsMD5, err := uploadutils.MD5FromGCS(ctx, s.bh, f)
+	if err == nil && gcsMD5 != "" {
+		// Skip if file exists.
+		return true, nil
+	}
+
+	ftpMD5, content, err := uploadutils.MD5FromHTTP(s.httpURLRoot + f)
+	if err != nil {
+		return false, fmt.Errorf("file %s cannot be downloaded from %s: %v", f, s.httpURLRoot, err)
+	}
+	if ftpMD5 == gcsMD5 {
+		// Skip if file exists.
+		return true, nil
+	}
+	log.Infof("Writing %s", f)
+	if resp, err := s.gc.FileUpload(ctx, &pb.FileRequest{
+		Filename: f,
+		Content:  content,
+		Md5Sum:   ftpMD5,
+		Project:  pb.FileRequest_ROUTEVIEWS,
+	}); err != nil {
+		return false, fmt.Errorf("FileUpload: err %v, resp %s", err, resp.String())
+	}
+	return false, nil
+}
+
+func (s *Synchronizer) uploadFilesFromFTP(ctx context.Context, files []string) {
+	var lastUploaded string
+	for _, f := range files {
+		err := backoff.Retry(func() error {
+			skipped, err := s.uploadFromFTP(ctx, f)
+			if err != nil {
+				return fmt.Errorf("uploadFromFTP(%s): %v", f, err)
+			}
+			if !skipped {
+				// Do not overwhelm the archive server.
+				time.Sleep(time.Second)
+			}
+			return nil
+		}, backoff.WithMaxRetries(backoff.NewConstantBackOff(s.retryInterval), s.retryCount))
+		if err != nil {
+			log.Error(err)
+			return
+		}
+
+		// Check if any archive is missing (two consecutive archives are over
+		// 30 minutes apart). Structured log for potential metric collection.
```

Comment:
you say 30 here, but check for 15 below (line 151)

Reply:
Done.